presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java

```diff
@@ -251,13 +269,45 @@ public static ParquetPageSource createParquetPageSource(
 
     public static org.apache.parquet.schema.Type getParquetType(HiveColumnHandle column, MessageType messageType, boolean useParquetColumnNames)
     {
+        org.apache.parquet.schema.Type type = null;
         if (useParquetColumnNames) {
-            return getParquetTypeByName(column.getName(), messageType);
+            type = getParquetTypeByName(column.getName(), messageType);
         }
 
         if (column.getHiveColumnIndex() < messageType.getFieldCount()) {
-            return messageType.getType(column.getHiveColumnIndex());
+            type = messageType.getType(column.getHiveColumnIndex());
         }
-        return null;
+
+        if (type == null) {
+            return null;
+        }
+
+        if (!schemaMatch(type, column)) {
+            throw new PrestoException(HIVE_PARTITION_SCHEMA_MISMATCH, format("The column %s is declared as type %s, but the Parquet file declares the column as type %s",
+                                            column.getName(),
+                                            column.getHiveType(),
+                                            type.asPrimitiveType().getPrimitiveTypeName()));
+        }
+        return type;
+    }
+
+    private static boolean schemaMatch(org.apache.parquet.schema.Type parquetType, HiveColumnHandle column)
+    {
+        String prestoType = column.getTypeSignature().getBase();
+        if (prestoType.equals(MAP) || prestoType.equals(ARRAY) || prestoType.equals(ROW)) {
```

Comment:
Could you please check the signature for the complex types recursively?

Reply:
I would suggest passing here the resolved `Type` object, so the convenient `Type#getTypeParameters` can be used to recurse into the nested types.

Reply:
e.g.: https://github.com/prestodb/presto/blob/e313922f234948bfba27f9334905a750580d824f/presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSource.java#L98